### PR TITLE
Adding kubernetes to guides draft 1

### PR DESCRIPTION
**Pull request by ethangj**

### CODE REVIEW

**Comment by ethangj** [comment]
@manualwise could you give this a readthrough? copy and paste from pdf was doing weird things and i'm not positive i could them all. @mlock as well.

**Comment by manualwise** [comment]
@ethangj mistakenly mentioned a non-Codeshipper here instead of @mlocher :)

**Comment by ethangj** [comment]
Oops. :(

File: _general/account/guides/deployment-with-kubernetes.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+In a nutshell, here's what this diagram is showing:
+
+- A Cluster is a collection of physical and/or virtual machines called Nodes.
+
+- Each Node is responsible for running a set of Pods.
+
+- A Pod is a group of networked Docker-based containers.
+
+Outside of the parent-child chain are **Deployments** (which I'll get to below) and **Services**. Services are logical sets of Pods with a de ned policy by which to access them (read: microservice). A service can span multiple Nodes within a Kubernetes Cluster.
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: _de ned_? Not sure if that was an issue with copy/paste or I just don't know the term.

File: _general/account/guides/deployment-with-kubernetes.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+## Deployments (uppercase) vs. deployments (lowercase)
+
+When it comes to actually launching containers, Kubernetes provides tools to automatically roll out new code by updating Deployment definitions. It is important to mention here that the word "Deployment" in Kubernetes-speak is really just a fancy (and a bit ambiguous) word for a recipe that describes how containers should be configured and launched. Because this article deals with delivering and launching updated Docker images to Kubernetes using Codeship, there
+is bound to be some confusion over terminology, so to keep things clear(ish) I'll be using the lowercase "deployment" to refer to the act of delivering product, and the uppercase "Deployment" to refer to the Kubernetes definition of the word.
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: please combine with the previous paragraph

File: _general/account/guides/deployment-with-kubernetes.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+I should point out that, even though Deployment updates need to be triggered in a specific way, there is very little risk of downtime in a multi-container environment. Thanks to the way Deployments are built, Kubernetes will ensure that no downtime is suffered by bringing down only a fraction of the Pods at a time. While the load won't necessarily be as efficiently distributed during these updates, the consumers of your application won't suffer any outages.
+
+## Integrating Codeship with Kubernetes
+
+So, given a functioning Kubernetes Deployment, how do we integrate it into our Codeship work now? The answer to this question ultimately depends on your Kubernetes host, but because the official documentation uses [Google Cloud](https://documentation.codeship.com/pro/continuous-deployment/google-cloud/) as an example, this is the platform I'll address. Thankfully for us, Codeship has already built out some Google Cloud integrations into their CI Platform for Docker that we can use to authenticate and deploy new images to Google Cloud.
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: _CI Platform for Docker_: should we refer to this as _Codeship Pro_?

File: _general/account/guides/deployment-with-kubernetes.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+- a Google Cloud Key – `GOOGLE_AUTH_JSON`
+- a Google Authentication Email – `GOOGLE_AUTH_EMAIL`
+- a Google Project ID – `GOOGLE_PROJECT_ID`
+
+Once we have an encrypted environment file (and have saved our Google Cloud environment variables to `gc.env.encrypted`), we next need to define the Google Cloud service in the `codeship-services.yml` file.
+
+```
+google_cloud_deployment:
+  image: zach ower/google-cloud-deployment
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: the image name is invalid. Also can we take a look and check the differences between this image and https://github.com/codeship-library/google-cloud-deployment and see if we can switch to our own?

File: _general/account/guides/deployment-with-kubernetes.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  add_docker: true
+  encrypted_env_ le: gc.env.encrypted
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: copy/paste error

File: _general/account/guides/deployment-with-kubernetes.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+  volumes:
+    - ./:/deploy
+
+gcr_dockercfg:
+  image: codeship/gcr-dockercfg-generator
+  add_docker: true
+  encrypted_env_ le: gc.env.encrypted
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: copy/paste error

File: _general/account/guides/deployment-with-kubernetes.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+```
+
+Notice that there are two services defined, rather than one. This is because one is for interacting with Google Cloud services ( `google_cloud_deployment` ), while another is used to enable Docker image push functionality to the Google Cloud Registry ( `gcr_dockercfg` ). This is only half of the puzzle, however, because although it creates the necessary services for interacting with Google Cloud, it doesn't automatically deploy newly built images or update a Kubernetes Deployment.
+
+## Google Container Registry Pushing
+
+Thanks to Codeship's built-in push steps, deploying
+a Docker image to a remote registry is a pretty painless process.
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: please combine with the previous paragraph

File: _general/account/guides/deployment-with-kubernetes.md [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+  registry: https://gcr.io
+  dockercfg_service: gcr_dockercfg
+```
+
+Now, when we push up our app image to the Google Container Registry, it will be tagged with the current build's Unix timestamp.
+
+## Updating Kubernetes Deployments
+
+Once our push step is defined, we need to tell Kubernetes to update the appropriate Deployment to roll out the new image. This is where the previously defined `google_cloud_deployment` service comes
+into play. Thanks to this service, we are able to easily run authenticated commands against Google Cloud Platform without any additional overhead, which means that manipulating our Kubernetes platform from within Codeship is no different than working with it directly.
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: please combine with the previous paragraph

File: _general/account/guides/deployment-with-kubernetes.md [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+Before we set up the Codeship step, though, let's take a look at how updating a Kubernetes Deployment actually works. According to the Kubernetes documentation (and as touched upon above), triggering a Deployment update is as simple as updating the Deployment's defined label or container image. For now, let's assume that we already have a defined Deployment for an Nginx server as per the documentation. All we have to do to roll out an updated Docker image to the Deployment is to change the de ned image using the kubectl command like so:
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: _de ned_ again. Not sure if that was an issue with copy/paste or I just don't know the term.

Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: Also please switch _kubectl_ to an inline code block `kubectl`

File: _general/account/guides/deployment-with-kubernetes.md [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+```
+$ kubectl set image deployment/nginx-deployment nginx=nginx:1.9.1 deployment "nginx-deployment" image updated
+```
+
+Because we went through the work of tagging our image pushes above, this type of update will be relatively easy for us to set up. But, this is just a command – It doesn't show us how to actually update Deployments from Codeship. All it takes to accomplish this is a small script to run the few necessary commands to authenticate to the Google Cloud Platform and trigger a Kubernetes Deployment update.
+
+Thankfully, due to the hard work put in by Codeship already, the script we need to write consists of only a small handful of commands:
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: I would rephrase this. I don't think our documentation is the right place to push how hard we're working :)

File: _general/account/guides/deployment-with-kubernetes.md [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+```
+#!/bin/bash 2
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: Do we know what the `2` is for?

Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: It's not in the repository version of this script at https://github.com/codeship/codeship-kubernetes-demo/blob/master/deploy.sh

File: _general/account/guides/deployment-with-kubernetes.md [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+Now that we have our deployment script set up (I've saved mine to the root of my project as `deploy.sh`), all we have left to do is to add a step to the `codeship-steps.yml` file that calls it:
+
+```
+- service: google_cloud_deployment
+  command: /deploy/deploy.sh
+```
+
+## Conclusion
+Fortunately for us, most of the heavy lifting for this integration has been done by Codeship already, which means that interacting with the Google Cloud Platform during our CI/CD process is as simple as running a command.
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: Would rephrase this :)

**Comment by ethangj** [comment]
@mlocher Updated, another check please!

File: _general/account/guides/deployment-with-kubernetes.md [hunk 13]
```diff
@@ -8,7 +8,7 @@ tags:
 category: Guides
 ---
 <div class="info-block">
-This was originally featured [as an eBook](https://resources.codeship.com/ebooks/deploy-docker-kubernetes-codeship-aw) in our [Resources Library](https://resources.codeship.com).
+This was originally featured [as an eBook](https://resources.codeship.com/ebooks/deploy-docker-kubernetes-codeship-aw) inyour[Resources Library](https://resources.codeship.com).
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: missing space between _in your_

File: _general/account/guides/deployment-with-kubernetes.md [hunk 14]
```diff
@@ -29,51 +29,49 @@ In a nutshell, here's what this diagram is showing:
...
 
 ## Integrating Codeship with Kubernetes
 
-So, given a functioning Kubernetes Deployment, how do we integrate it into our Codeship work now? The answer to this question ultimately depends on your Kubernetes host, but because the official documentation uses [Google Cloud](https://documentation.codeship.com/pro/continuous-deployment/google-cloud/) as an example, this is the platform I'll address. Thankfully for us, Codeship has already built out some Google Cloud integrations into their CI Platform for Docker that we can use to authenticate and deploy new images to Google Cloud.
+So, given a functioning Kubernetes Deployment, how do we integrate it intoyourCodeship work now? The answer to this question ultimately depends on your Kubernetes host, but because the official documentation uses [Google Cloud](https://documentation.codeship.com/pro/continuous-deployment/google-cloud/) as an example, this is the platform we'll address. To make this easier we've built out some Google Cloud integrations into Codeship Pro that we can use to authenticate and deploy new images to Google Cloud.
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: missing spaces _in to your Codeship_

File: _general/account/guides/deployment-with-kubernetes.md [hunk 15]
```diff
@@ -29,51 +29,49 @@ In a nutshell, here's what this diagram is showing:
...
 
-Before we can do anything, however, we need to create an encrypted environment file using Codeship's Jet CLI tool in order to authenticate to Google Cloud. Codeship already has an [excellent tutorial](https://documentation.codeship.com/pro/getting-started/encryption/) of how to do this, so I won't go over it here, but the environment variables that need to be set are:
+Before we can do anything, however, we need to create an encrypted environment fileyour[Jet CLI]({% link _pro/getting-started/installation.md %}) tool in order to authenticate to Google Cloud. We have a [documentation article](https://documentation.codeship.com/pro/getting-started/encryption/) on how to do this, so we won't go over it here - but the environment variables that need to be set are:
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: missing spaces _file your Jet_

File: _general/account/guides/deployment-with-kubernetes.md [hunk 16]
```diff
@@ -29,51 +29,49 @@ In a nutshell, here's what this diagram is showing:
...
 
 - a Google Cloud Key – `GOOGLE_AUTH_JSON`
 - a Google Authentication Email – `GOOGLE_AUTH_EMAIL`
 - a Google Project ID – `GOOGLE_PROJECT_ID`
 
-Once we have an encrypted environment file (and have saved our Google Cloud environment variables to `gc.env.encrypted`), we next need to define the Google Cloud service in the `codeship-services.yml` file.
+Once we have an encrypted environment file (and have savedyourGoogle Cloud environment variables to `gc.env.encrypted`), we next need to define the Google Cloud service in the `codeship-services.yml` file.
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: missing spaces _saved your Google_

File: _general/account/guides/deployment-with-kubernetes.md [hunk 17]
```diff
@@ -29,51 +29,49 @@ In a nutshell, here's what this diagram is showing:
...
+  encrypted_env_file: gc.env.encrypted
 ```
 
 Notice that there are two services defined, rather than one. This is because one is for interacting with Google Cloud services ( `google_cloud_deployment` ), while another is used to enable Docker image push functionality to the Google Cloud Registry ( `gcr_dockercfg` ). This is only half of the puzzle, however, because although it creates the necessary services for interacting with Google Cloud, it doesn't automatically deploy newly built images or update a Kubernetes Deployment.
 
 ## Google Container Registry Pushing
 
-Thanks to Codeship's built-in push steps, deploying
-a Docker image to a remote registry is a pretty painless process.
+Your `codeship-steps.yml` file supports `push` steps, to make pushing Docker images to a remote registry more efficient. We'll use a `push` step to push toyourGoogle Container Registry accout.
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: missing spaces _to your Google_

File: _general/account/guides/deployment-with-kubernetes.md [hunk 18]
```diff
@@ -29,51 +29,49 @@ In a nutshell, here's what this diagram is showing:
...
 
-Using the `gcr_dockercfg` service defined above, all we need to do is add a step to the `codeship-steps.yml` file with our Google Container Registry URL as the destination. It's important to remember here that we will be deploying our application image, so be sure to replace the app service name with the name of the service your own application is running on.
+Using the `gcr_dockercfg` service defined above, all we need to do is add a step to the `codeship-steps.yml` file withyourGoogle Container Registry URL as the destination. It's important to remember here that we will be deployingyourapplication image, so be sure to replace the app service name with the name of the service your own application is running on.
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: missing spaces _with your Google_ and _deploying your application_

File: _general/account/guides/deployment-with-kubernetes.md [hunk 19]
```diff
@@ -135,26 +132,23 @@ deployment-name app=gcr.io/project-name/app-name:$CI_TIMESTAMP
 
 You can find the complete repository including the `codeship-steps.yml` and the `codeship-services.yml` files here: [https://github.com/codeship/codeship-kubernetes-demo](https://github.com/codeship/codeship-kubernetes-demo)
 
-Let's step through the above script really quick. The first important command is the authentication piece. The `google_cloud_deployment` service needs to be authenticated with the Google Cloud Platform before we can run any commands. Since we set up the necessary environment variables already, all it takes to authenticate is to run the `codeship_google authenticate` command at the beginning of our script.
+Let's step through the above script really quick. The first important command is the authentication piece. The `google_cloud_deployment` service needs to be authenticated with the Google Cloud Platform before we can run any commands. Since we set up the necessary environment variables already, all it takes to authenticate is to run the `codeship_google authenticate` command at the beginning ofyourscript.
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: missing spaces _of your script_

File: _general/account/guides/deployment-with-kubernetes.md [hunk 20]
```diff
@@ -135,26 +132,23 @@ deployment-name app=gcr.io/project-name/app-name:$CI_TIMESTAMP
...
 
 Next, we need to set the compute zone. This example shows `us-central1-a`, but you should change this to suit your needs. The next set of commands is the actual Kubernetes interactions. The first sets the Kubernetes cluster that we need to interact with, while the second is the actual Deployment update command. As you can see, it's not very different from the example provided by Kubernetes itself.
 
 It's important to note here that Codeship provides an environment variable of the current build's timestamp, which allows us to correlate the Kubernetes command with the registry push step above.
 
-Now that we have our deployment script set up (I've saved mine to the root of my project as `deploy.sh`), all we have left to do is to add a step to the `codeship-steps.yml` file that calls it:
+Now that we haveyourdeployment script set up (I've saved mine to the root of my project as `deploy.sh`), all we havefileft to do is to add a step to the `codeship-steps.yml` file that calls it:
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: missing spaces _have your deployment_ and _have left_ (there is also an additional _fi_ in the last one)

File: _general/account/guides/deployment-with-kubernetes.md [hunk 21]
```diff
@@ -135,26 +132,23 @@ deployment-name app=gcr.io/project-name/app-name:$CI_TIMESTAMP
...
-
-Thanks to the incredible flexibility of Codeship, interacting with any cloud platform that we choose is an incredibly straightforward process. Because we are only limited by the capabilities of Docker itself, our deployment workflows are completely customizable, so we can get our process just right.
+## To Learn More
 
 [Codeship Pro]({% link _pro/getting-started/getting-started.md %})
 
 ![Slack Screenshot]({{ site.baseurl }}/images/jenkins-guide/slack_screenshot.png)
 
-If you want to learn more please contact us at [helpdesk@codeship.com](mailto:helpdesk@codeship.com).
+If you want tofilearn more please contact us at [helpdesk@codeship.com](mailto:helpdesk@codeship.com).
```
Comment (mlocher) [inline _general/account/guides/deployment-with-kubernetes.md]: missing space (but added fi) on _to learn_
